### DISCUSSION TITLE
Widget Plugin/deephaven.ui component Ideas

### DISCUSSION THREAD

**Discussion by mofojed** [deephaven/web-client-ui#2022] [Ideas]
What kind of plugins would you like to see built in Deephaven? Use a WidgetPlugin as a starting point: https://github.com/deephaven/deephaven-plugins/tree/main/templates

Or just build a component in deephaven.ui: https://github.com/deephaven/deephaven-plugins/tree/main/plugins/ui/docs

**Comment by mofojed** [comment]
A PivotTable widget would be great.

**Reply by rbasralian** [reply]
This would be fantastic, especially one that is as flexible and easy to create like the Swing pivot widget https://deephaven.io/enterprise/docs/interfaces/classic/pivot-widgets/#build-a-pivot-widget

**Comment by mofojed** [comment]
Copilot or Chat GPT bot for asking questions.

**Comment by mofojed** [comment]
Voice chat AI assistant

**Comment by mofojed** [comment]
Speedometer gauge widget, maybe something like https://www.npmjs.com/package/react-d3-speedometer ?

**Comment by mattrunyon** [comment]
Audio notification widget or component. Like an alarm

**Comment by mattrunyon** [comment]
A clock could be useful. Maybe some sort of timer/stopwatch too

**Comment by mattrunyon** [comment]
UGP viewer for a specific table

**Comment by AkshatJawne** [comment]
A currency exchange rate table, with live exchange rates of popular currencies.

**Comment by AkshatJawne** [comment]
A little market news table, get live info about specific market sector or stock prices?

**Comment by AkshatJawne** [comment]
A way to schedule meaningful events, a small calendar-type thing?

**Comment by mofojed** [comment]
IPython widget display

**Comment by AkshatJawne** [comment]
Spotify / Apple Music widget could be nice, not sure about practicality, but would be fun to be able to work with data and switch songs in the UI itself 😄

**Comment by mattrunyon** [comment]
A floating component for dh.ui. Like the linker modal, but something you can drag around and overlays everything. Could be useful for something like a floating control panel instead of needing to stick controls in a panel.

**Comment by mofojed** [comment]
Tetris

**Comment by JamesXNelson** [comment]
A voting widget, so we can collect the next set of hackathon ideas in-house ^-^

**Comment by JamesXNelson** [comment]
A poker game that count cards for you and plots your % chance to win in realtime :smiling_imp:

**Comment by devinrsmith** [comment]
Flip card; a two sided component. Typically I might expect plot in front, and you can flip it over to learn more... maybe backside has markdown, code used to create it, or anything else.

**Reply by mattrunyon** [reply]
This is just a proof of concept, but purely in dh.ui. No other plugins required

```py
from deephaven import ui

styles = """
.flip-card {
  background-color: var(--dh-color-surface-bg);
  height: 100%;
  width: 100%;
  flex-shrink: 1;
  perspective: 1000px; /* Remove this if you don't want the 3D effect */
}

.flipping {
    overflow: hidden;
}

.flip-card-inner {
  position: relative;
  background-color: var(--dh-color-surface-bg);
  width: 100%;
  height: 100%;
  transition: transform 0.8s;
  transform-style: preserve-3d;
}

/* Do an horizontal flip when you move the mouse over the flip box container */
.flip-card.flipped .flip-card-inner {
  transform: rotateY(180deg);
}

/* Position the front and back side */
.flip-card-front, .flip-card-back {
  position: absolute;
  width: 100%;
  height: 100%;
  -webkit-backface-visibility: hidden; /* Safari */
  backface-visibility: hidden;
}

/* Style the front side (fallback if image is missing) */
.flip-card-front {
}

/* Style the back side */
.flip-card-back {
  transform: rotateY(180deg);
}

.flip-btn {
    position: absolute;
    z-index: 1;
    top: 0.5rem;
    left: initial;
    right: 0.5rem;
}
"""

@ui.component
def flip_card(front, back):
    flipped, set_flipped = ui.use_state(False)
    flipping, set_flipping = ui.use_state(False)

    def handle_flip():
        set_flipped(lambda f: not f)
        set_flipping(True)

    flip_btn = ui.action_button(
        ui.icon('vsSync'),
        on_press=handle_flip,
        UNSAFE_class_name='flip-btn'
    )

    return [
        ui.html.style(styles),
        ui.html.div(
            ui.html.div(
                ui.html.div(
                    flip_btn,
                    front,
                    className='flip-card-front'
                ),
                ui.html.div(
                    flip_btn,
                    back,
                    className='flip-card-back'
                ),
                className='flip-card-inner'
            ),
            className='flip-card' + (' flipped' if flipped else '') + (' flipping' if flipping else ''),
            on_transition_end=lambda: set_flipping(False)
        )
    ]

from deephaven.plot import express as dx

_stocks = dx.data.stocks()
_plot = dx.line(_stocks, x="Timestamp", y="Price", by="Sym")

a = flip_card(_stocks, _plot)
```

**Reply by mofojed** [reply]
It's also in a Gist: https://gist.github.com/mattrunyon/cb60e1b247ec439dd7ec6997c36b5693

**Comment by bmingles** [comment]
Anything to help communicating across time zones + units of measure in conversations with team members.

Maybe show localized stats for team members. Things like:
- Current temperature in Farenheit and Celcius
- Current time in their timezone

**Comment by chipkent** [comment]
Git widget

**Reply by mofojed** [reply]
@chipkent can you clarify what type of git operations you'd be looking for?

**Comment by chipkent** [comment]
My biggest complaint about the web ui is that it feels like a toy compared to some swing things.  In particular, it is limited to 1 level deep, while swing allowed more component nesting.  I'd like components that support deeper nesting or other navigation to handle complex scenarios.  This may already be covered.

**Comment by chipkent** [comment]
Video widget.  For the video case, a live camera feed could be displayed.  This would let the visual data be shown in parallel with DH code that is analyzing it. and video widgets.  For the video case, a live camera feed could be displayed.  This would let the visual data be shown in parallel with DH code that is analyzing it.

**Reply by mofojed** [reply]
Would the video widget take a local live feed? Or just provide a link to a live feed?

**Comment by chipkent** [comment]
Image widget.  This would be useful for creating UIs for labeling data for AI -- among other things.

**Reply by mofojed** [reply]
Ethan is actually working on this currently: https://github.com/deephaven/deephaven-plugins/issues/597

**Comment by chipkent** [comment]
Burndown chart

**Reply by mofojed** [reply]
We should be able to build this with existing components: https://en.wikipedia.org/wiki/Burndown_chart
Providing a component that easily wraps this would be nice.

**Comment by devinrsmith** [comment]
Iceberg catalog navigation (similar in spirit to DHE `db`, where tables have a namespace and table name)

**Comment by rachelmbrubaker** [comment]
Map / geographical display for data.

**Reply by devinrsmith** [reply]
https://www.nyc.gov/site/tlc/about/tlc-trip-record-data.page shapefiles would be a good example to use for a smaller / local geographic area.

**Reply by mofojed** [reply]
We do have support for maps/geographic data through plotly: 
![image](https://github.com/user-attachments/assets/3fc7b6cf-aaac-4d84-bf00-48e238734324)

We haven't fully documented it yet though. A few examples from the PR that merged them: https://github.com/deephaven/deephaven-plugins/pull/71#issuecomment-1739931779

**Comment by mofojed** [comment]
An Olympic event dashboard. Possible API: https://olympics.pressassociation.io/docs/coverage

**Comment by mofojed** [comment]
Tic-tac-toe ui.component, like the React tutorial: https://react.dev/learn/tutorial-tic-tac-toe

**Comment by mofojed** [comment]
Email notifcation

**Comment by mofojed** [comment]
Log panel

**Comment by devinrsmith** [comment]
Chat window (Colin had a JS client version of this a while ago I think).

**Comment by mofojed** [comment]
File upload widget

**Comment by devinrsmith** [comment]
Submit / Form builder (produces into TablePublisher)

**Reply by mofojed** [reply]
You can do this with `deephaven.ui` using `ui.form`

**Reply by devinrsmith** [reply]
Sure; but this would be a helper to build it based on a Table definition probably, and do all the wiring up for you.

**Reply by mofojed** [reply]
Here's an example wiring up a Form with a publish automatically: https://gist.github.com/mofojed/ef33d723c0846f2bcdf7575cd2967883
Note there are some improvements still that can be made:
- Only wired up for number and string types
- Need a proper decimal number_field input: https://github.com/deephaven/deephaven-plugins/issues/695
- Form can allow validation errors to be passed in to inform the user of erroneous input: https://github.com/deephaven/deephaven-plugins/issues/694

**Comment by bmingles** [comment]
Configurable action buttons to run arbitrary scripts or ones in paths within notebooks folder.

**Comment by mofojed** [comment]
Location widget. Default to users location or allow them to manually enter one.
